docs_src/design/adr/security/0027-authentication.md

```diff
@@ -0,0 +1,257 @@
+  socket with token-based authentication,
+  with the token being mediated by Hashicorp Vault.
+  An access control list limits access to the keyspace of
+  the configuration store.
+
+* Communication with EdgeX's default database, Redis, is secured using
+  username/password authentication, with the password stored
+  in Hashicorp Vault.  An access control list limits the commands
+  that clients are allowed to issue to the server.
+
+External access to EdgeX microservices has also been secured.
+EdgeX microservices only bind to local ports,
+and are only exposed externally through a Kong API gateway.
+This gateway is configured to use TLS 1.3,
+using JWT authentication using public key cryptography.
+All external requests are filtered at the API gateway.
+URL rewriting is used to concentrate microservices
+on a single HTTP-accessible port.
+
+Behind the proxy, it is not possible to verify Kong
+as the origin of local network traffic because mutual-auth TLS
+is not supported in the open source version of Kong.
+Although the Kong JWT plugin will set request headers
+on the backend request that identify the caller,
+there is no mechanism by which Kong can prove to a
+backend service that it was the component that
+performed the authentication step.
+Even though the original JWT passes through the proxy,
+the Kong authentication plugins do not expose
+token introspection endpoints that the backend service
+could use to check token validity independently.
+
+The consequence of having an API gateway that performs
+all microservice authentication is that communication
+between EdgeX microservices running behind the API gateway
+are not authenticated in any way.
+EdgeX microservices are unable to distinguish
+malicious traffic that has evaded the API gateway
+from legitimate microservice traffic.
+
+## Proposed Design
+
+This ADR proposes an implementation of the
+[Microservice Authentication UCR](../../ucr/Microservice-Authentication.md)
+that uses a token-based authentication mechanism.
+
+This ADR proposes to relieve the Kong API gateway of its
+[JWT](https://www.rfc-editor.org/rfc/rfc7519) management responsibility,
+and instead use Hashicorp Vault for this purpose,
+which is already used as EdgeX's secret store.
+In turn, EdgeX can use a simpler reverse proxy, NGINX,
+resulting in a total savings of 150MB of memory
+and 300MB of storage (in the docker implementation).
+This change requires minimal modification of existing
+clients written to perform JWT-based authentication at the Kong gateway:
+they simply use a Vault-issued JWT
+instead of a Kong-issued JWT or a self-issued JWT.
+
+This ADR proposes a layered authentication scheme,
+with the reverse proxy performing an initial check for all external requests,
+and EdgeX services themselves authenticating all internal and external requests.
+The layered approach is required,
+as some EdgeX endpoints such as `/api/v2/ping`
+must be anonymous for health-checking purposes.
```

Comment:
I thought I did a pretty good job at explaining it.  Not all of the URLs can presently have security enabled, like api/v2/ping.  I also mention programmer mistakes elsewhere, if I am not mistaken.  Do you have suggested alternative wording?

Reply:
I am wondering why we need the initial check at all. If services authenticate requests (second layer), why do we need to authenticate in the proxy (first layer) too? If my assumption about security and performance considerations is correct, then I suggest having that explanation in the ADR.

Reply:
Longer explanation:

This ADR proposes a layered authentication scheme,
with the reverse proxy performing an initial check for all external requests,
and EdgeX services themselves authenticating all internal and external requests.
There are three reasons for the layered approach:

1. Authentication at the proxy layer provides a choke point
   and policy enforcement points for incoming requests.
   By customizing the behavior of the proxy-auth component,
   it is possible to allow access to some URLs
   and deny access to other URLs based on arbitrary criteria,
   such as source IP address, JWT-based claims,
   or user identity and role mappings.

2. It means that individual microservices do not immediately
   need to implement fine-grained authorization to get the
   same effect as having custom policy enforcement at the proxy.

3. It provides defense-in-depth against microservice implementation bugs
   and other technical debt that might otherwise put EdgeX
   microservices at risk.  Getting a known response to
   `/core-data/api/v2/ping` as a result of an anonymous HTTP request
   would positively identify an EdgeX installation.
   Similarly, an adopter porting their custom services to EdgeX 3.0
   without adding authentication hooks could be vulnerable to outside attacks
   that might be mitigated by the additional check at the proxy layer.